Use the new swap chain model of gfx-hal

zakorgy opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/szeged/webrender/331)
<!-- Reviewable:end -->

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -847,9 +868,6 @@ impl<B: hal::Backend> Device<B> {
         self.device.wait_idle().unwrap();
 
         let ref mut heaps = *self.heaps.lock().unwrap();
-        for frame in self.frames.drain(..) {
```
we still want to drain this, I think

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1007,40 +1016,73 @@ impl<B: hal::Backend> Device<B> {
-        let (swap_chain, images) =
-            unsafe { device.create_swapchain(surface, swap_config, old_swap_chain) }
-                .expect("create_swapchain failed");
+        println!("## Image count {}", frame_count);
```
debug print?

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1007,40 +1016,73 @@ impl<B: hal::Backend> Device<B> {
+        frame_depths.push(depth);
+        for _ in 1 .. frame_count {
+            frame_depths.push(
+                DepthBuffer::new(
```
why are we creating N depth buffers? we should be able to use the same depth buffer for all frames
moreover, with picture caching we may not even need any depth buffers for the swapchain images

zakorgy replied:
As I've found picture caching uses the depth buffer.

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1007,40 +1016,73 @@ impl<B: hal::Backend> Device<B> {
+        }
 
-        info!("Frames: {:?}", frames);
+        println!("## Frame depth count {:?}", frame_depths.len());
```
debug print

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -2208,11 +2250,13 @@ impl<B: hal::Backend> Device<B> {
             (img.format, &img.core, fbo.layer_index)
         } else {
-            (
+            println!("## This path is unimplemented!");
```
`warn` or `error`?

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -3254,36 +3302,39 @@ impl<B: hal::Backend> Device<B> {
+                                self.viewport.rect,
+                                &self.render_passes,
+                            );
+                            let old_frame = mem::replace(&mut self.frame, new_frame);
```
should we just do `self.frame.surface_image = Some(surface_image))` instead of creating a new one and doing the `mem::replace`?

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -3311,41 +3363,34 @@ impl<B: hal::Backend> Device<B> {
         unsafe {
-            match self.swap_chain.as_mut() {
-                Some(swap_chain) => {
+            let present_pass = self.render_passes.present_pass();
```
why are we doing this pass? it seems empty

zakorgy replied:
This is just a workaround to fix validation layer errors (that the frame image is not in a presentable layout) until I write the code which identifies the last render pass related to the frame image and we can start this pass there.

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -171,16 +172,54 @@ struct Fence<B: hal::Backend> {
+
+    fn deinit(self, device: &B::Device) {
         unsafe {
             device.destroy_framebuffer(self.framebuffer);
```
We need to be destroying the associated frame buffer *right after* it was presented.
See the quad example for the reference. If keep it, we are forcing Metal to have more than one `CAMetalDrawable` in flight.
Currently, the code appears to be destroying the framebuffer only after the next acquisition is done.

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1007,40 +1016,61 @@ impl<B: hal::Backend> Device<B> {
+        };
+
+        let surface_image = unsafe {
+            match surface.acquire_image(!0) {
```
why are we acquiring an image in more than one place in the code?

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -3311,41 +3349,34 @@ impl<B: hal::Backend> Device<B> {
-            match self.swap_chain.as_mut() {
-                Some(swap_chain) => {
+            let present_pass = self.render_passes.present_pass(self.frame_depth.is_some());
+            self.command_buffer.begin_render_pass(
```
we definitely not want to have this pass :)

zakorgy commented:
A quick update on this PR: it's is in a state where everything works without validation layer errors with Vulkan backend. But I still need to nail down the number of render passes we use for the main target.
Changes since the last update:
- Enabled read pixels functionality again
- Enabled headless mode again
- Finally fixed the validation layer errors during resize ( https://github.com/szeged/webrender/issues/80#issuecomment-488642315 ): turned out the issue was that we handled the window dimensions and the viewport within the same variable.
